### DISCUSSION TITLE
How do you access the asset folder in an unpackages application?

### DISCUSSION THREAD

**Discussion by algjohnston** [microsoft/WindowsAppSDK#3175] [Q&A]
I added this to the csproj file
`
  <ItemGroup>
	<None Update="Assets/*">
		<CopyToOutputDirectory>Always</CopyToOutputDirectory>
	</None>
  </ItemGroup>
`
so I can access the asset folder after an install.
However, I can not find any documentation or help on how to access these files.

I found this page
https://learn.microsoft.com/en-us/uwp/api/windows.storage.storagefolder.getfolderfrompathasync?view=winrt-22621
, but when I try it out, this line throws an InvalidOperationException
`
            string root = Windows.ApplicationModel.Package.Current.InstalledLocation.Path;
`

**Comment by ghost1372** [comment]
Try this

```CS
public static string GetFullPathToExe()
 { 
var path = AppDomain.CurrentDomain.BaseDirectory;
 var pos = path.LastIndexOf("\\"); 
return path.Substring(0, pos); 
} 
public static string GetFullPathToAsset(string assetName) 
{ 
return GetFullPathToExe() + "\\Assets\\" + assetName; 
}

```

**Reply by DrusTheAxe** [reply]
>Windows.ApplicationModel.Package.Current

If you're an unpackaged app `Package.Current` fails (`APPMODEL_ERROR_NO_PACKAGE`) because your process lacks package identity. As expected.

If you're an unpackaged app then you can use [GetModuleFileName](https://learn.microsoft.com/windows/win32/api/libloaderapi/nf-libloaderapi-getmodulefilenamew)() to retrieve your process' executable path. See GetModulePath() in [test\inc\WindowsAppRuntime.Test.FileSystem.h](https://github.com/microsoft/WindowsAppSDK/blob/main/test/inc/WindowsAppRuntime.Test.FileSystem.h) e.g.

```c++
std::filesystem::path path{ GetModulePath(nullptr) };
path /= L"Assets";
const auto filename{ path / L"foo.bar" };
```

**Comment by DarranRowe** [comment]
If you are adding these files to the project, are these being added to the resources.pri file?

If they are, you can easily query for a resource file, and it will provide a full path to that file. As an example, I have a project that had a bunch of png files added. If I dump the contents of the resource index file, using the command line makepri dump /if resources.pri, then the generated file contains entries such as:

```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<PriInfo>
	<ResourceMap name="TestApp" version="1.0" primary="true">
		<Qualifiers/>
		<ResourceMapSubtree name="Files">
			<!--xaml xbf files here-->
			<ResourceMapSubtree name="Assets">
				<NamedResource name="01.png" uri="ms-resource://TestApp/Files/Assets/01.png">
					<Candidate type="Path">
						<Value>Assets\01.png</Value>
					</Candidate>
				</NamedResource>
				<!--other png files here-->
			</ResourceMapSubtree>
		</ResourceMapSubtree>
	</ResourceMap>
</PriInfo>
```

As you should see, this does give a path to the files added to the project.

If I use the resource manager API to query for this:

```
//This gets the path to resources.pri
auto default_resource_name = winrt::Microsoft::Windows::ApplicationModel::Resources::ResourceLoader::GetDefaultResourceFilePath();
auto resource_manager = winrt::Microsoft::Windows::ApplicationModel::Resources::ResourceManager(default_resource_name);
auto resource_map = resource_manager.MainResourceMap();
auto resource = resource_map.GetValue(L"Files/Assets/01.png");

resourcePath().Text(resource.ValueAsString());
```

Then the value that I get in the text block that I named resourcePath is:

![Screenshot 2022-11-23 140243](https://user-images.githubusercontent.com/52577874/203566401-febebea6-56af-4ca8-b0f9-4b69cf4ad3a4.png)

The resources API calculates the path for you. While this is C++, the same method should work for C# too.

The only issue is that the unpackaged project builds don't always create resources.pri. I think this normally gets generated as part of the packaging process. It is easy enough to deal with this, if you only have one .pri file then just copy that to resources.pri. If you are using components that also have .pri files, then you would have to merge them or access each .pri file.

If there is no .pri files, then I agree with ghost1372's suggestion of getting the base application directory and working from there.
